Remove included calendars from gem (#54)
gocardless/business: README.md

## Getting started

To install business, simply:

```bash
gem install business
```

If you are using a Gemfile:

```ruby
gem "business", "~> 2.0"
```

### Creating a calendar

Get started with business by creating an instance of the calendar class, passing in a hash that specifies with days of the week are considered working days, and which days are holidays.

```ruby
calendar = Business::Calendar.new(
  working_days: %w( mon tue wed thu fri ),
  holidays: ["01/01/2014", "03/01/2014"]    # array items are either parseable date strings, or real Date objects
  extra_working_dates: [nil], # Makes the calendar to consider a weekend day as a working day.
)
```

### Use a calendar file

Defining a calendar as a Ruby object may not be convenient, so we provide a way of defining these calendars as YAML. Below we will walk through the necessary [steps](#example) to build your first calendar. All keys are optional and will default to the following:

- If `working_days` is missing, then common default is used (mon-fri).
- If `holidays` is missing, "no holidays" assumed.
- If `extra_working_dates` is missing, then no changes in `working_days` will happen.

> > Note: Elements of `holidays` and `extra_working_dates` may be either strings that `Date.parse()` can understand, or YYYY-MM-DD (which is considered as a Date by Ruby YAML itself).

Author: I'll add a documentation link for `Date.parse()`

#### Example

```yaml
# lib/calendars/my_calendar.yml
working_days:
  - Monday
  - Wednesday
  - Friday
holidays:
  - 1st April 2020
  - 2021-04-01
extra_working_dates:
  - 9th March 2020 # A Saturday
```

> Ensure the calendar file is saved to a directory that will hold all your calendars, eg; `lib/calendars` then add this directory to `load_paths` before you call your calendar. `load_paths` can also accept an array of hashes, keyed with the calendar's name.

Reviewer: ```suggestion
Ensure the calendar file is saved to a directory that will hold all your calendars, eg; `lib/calendars` then add this directory to `load_paths` before you call your calendar. 

`load_paths` also accepts a plain Ruby has with the format:
`{ "< calendar_name >" => { "< day_type >" => ["< value >"] }`
```

```ruby
Business::Calendar.load_paths = [
  "lib/calendars",
  { "foo_calendar" => { "working_days" => ["monday"] } },
]
```

> Now you can load the calendar by calling the `Business::Calendar.load(calendar_name)`. The
> `Business::Calendar.load_cached(calendar_name)` variant of this method caches the calendars by name after loading them, to avoid reading and parsing the config file multiple times.

Reviewer: ```suggestion
Now you can load the calendar by calling the `Business::Calendar.load(calendar_name)`. 

In order to avoid parsing the calendar file multiple times, there is a `Business::Calendar.load_cached(calendar_name)` method that caches the calendars by name after loading them.
```

Date calculations based on business calendars.

- [v2.0.0 breaking changes](#v200-breaking-changes)
- [Getting Started](#getting-started)
  - [Creating a calendar](#creating-a-calendar)
  - [Using a calendar file](#use-a-calendar-file)
- [Checking for business days](#checking-for-business-days)
- [Business day arithmetic](#business-day-arithmetic)
- [But other libraries already do this](#but-other-libraries-already-do-this)
- [License & Contributing](#license--contributing)

## v2.0.0 breaking changes

We have removed the bundled calendars as of version 2.0.0, if you need the calendars that were included:

> - Download the calendars you wish to use from [before version 2](https://github.com/gocardless/business/tree/b12c186ca6fd4ffdac85175742ff7e4d0a705ef4/lib/business/data)

Reviewer: ```suggestion
- Download the calendars you wish to use from [v1.18.0](https://github.com/gocardless/business/tree/b12c186ca6fd4ffdac85175742ff7e4d0a705ef4/lib/business/data)
```

> - Place them in a suitable place in your project, eg: `lib/calendars`

Reviewer: ```suggestion
- Place them in a suitable directory in your project, typically `lib/calendars`
```

> - Add the directory to where you placed the yml files before you load the calendar

Reviewer: ```suggestion
- Add this directory path to your instance of `Business::Calendar` using the `load_paths` method.
```

```ruby

> Calendar::Business.load_paths("lib/calendars") # your_project/lib/calendars/ contains bacs.yml
> Calendar::Business.load("bacs")

Reviewer: ```suggestion
Business::Calendar.load_paths("lib/calendars") # your_project/lib/calendars/ contains bacs.yml
Business::Calendar.load("bacs")
```

)
```

### Use a calendar file

Defining a calendar as a Ruby object may not be convenient, so we provide a way of defining these calendars as YAML. Below we will walk through the necessary [steps](#example-calendar) to build your first calendar. All keys are optional and will default to the following:

Note: Elements of `holidays` and `extra_working_dates` may be either strings that `Date.parse()` [can understand](https://ruby-doc.org/stdlib-2.7.1/libdoc/date/rdoc/Date.html#method-c-parse), or `YYYY-MM-DD` (which is considered as a Date by Ruby YAML itself)[https://github.com/ruby/psych/blob/6ec6e475e8afcf7868b0407fc08014aed886ecf1/lib/psych/scalar_scanner.rb#L60].

#### YAML file Structure

```yml
working_days: # Optional, default [Monday-Friday]
  -
holidays: # Optional, default: []  ie: "no holidays" assumed
  -
extra_working_dates: # Optional, default: [], ie: no changes in `working_days` will happen
  -
```

#### Example calendar

```yaml
# lib/calendars/my_calendar.yml
working_days:
  - Monday
  - Wednesday
  - Friday
holidays:
  - 1st April 2020
  - 2021-04-01
extra_working_dates:
  - 9th March 2020 # A Saturday
```

> Ensure the calendar file is saved to a directory that will hold all your calendars, eg; `lib/calendars` then add this directory to `load_paths` before you call your calendar.

Reviewer: ```suggestion
Ensure the calendar file is saved to a directory that will hold all your calendars, typically `lib/calendars`, then add this directory to your instance of `Business::Calendar` using the `load_paths` method before you call your calendar.
```

$ gem install business
```

> ## Important: 2.0.0 breaking changes

Reviewer: Could we add a perm link to the `v1.18.0` branch as a courtesy to those who wouldn't want to upgrade?

Reviewer 2: wouldn't they just pin the version in their `Gemfile` to v.1.18.0?

Reviewer: True, but theres also a courtesy part where you link to the branch on the README as the documentation would be different.

We have removed the bundled calendars as of version 2.0.0, if you need the calendars that were included:

> Download the calendars you wish to use from [before version 2](https://github.com/gocardless/business/tree/b12c186ca6fd4ffdac85175742ff7e4d0a705ef4/lib/business/data) and place them in a suitable place in your project.

Reviewer: Are we able to provide a perm link to the old data? Concerned that the sha in the link may shift and result in a 404.

Reviewer 2: The SHA Is how github generates permanent links. One of the guarantees Git gives you is that the SHA will never shift and Github just exposes that.

Author: This is the permalink, press `y` on any file you're viewing to expand the link to a permalink

Then, add the directory to where you placed the yml files before you load the calendar:

```ruby
Calendar::Business.load_paths("lib/calendars") # your_project/lib/calendars/ contains bacs.yml
Calendar::Business.load("bacs")

> ```

Reviewer: Could we normalise on `lib/calendars` being the "suitable place" in the project?

Reduces the overhead and will allow us to reword this whole section to be clearer

Author: Good shout

### Getting started

Get started with business by creating an instance of the calendar class, passing in a hash that specifies with days of the week are considered working days, and which days are holidays.

```ruby
calendar = Business::Calendar.new(
  working_days: %w( mon tue wed thu fri ),
  holidays: ["01/01/2014", "03/01/2014"]    # array items are either parseable date strings, or real Date objects
  extra_working_dates: [nil], # Makes the calendar to consider a weekend day as a working day.
)
```

### Load a calendar from a file

#### Calendar file definition

> Defining a calendar as a Ruby object may not be convient, to load it from a YAML file follow and customise the example below. All keys are optional and will default to the following:

Reviewer: ```suggestion
Defining a calendar as a Ruby object may not be convenient, `business` also allows you to define a calendar as a YAML file and load this into your application.

Below we will walk through the necessary steps to build your first calendar.
```

Can we follow the thread of this and walk through the creation of a calendar?

Author: 👍
